export_rsvps.py

```diff
@@ -0,0 +1,84 @@
+import hashlib
+
+import psycopg2
+import psycopg2.extras
+from pywell.entry_points import run_from_cli, run_from_api_gateway
+
+import validate_key
+
+DESCRIPTION = 'Download RSVPs.'
+
+ARG_DEFINITIONS = {
+    'DB_HOST': 'Host for PostgreSQL connection.',
+    'DB_NAME': 'Database name.',
+    'DB_PASS': 'Pass for database connection.',
+    'DB_PORT': 'Port for database connection.',
+    'DB_SCHEMA': 'Scheam for database query.',
+    'DB_USER': 'Username for database connection.',
+    'EXTRA_WHERE': 'Anything to add to the WHERE clause in the RSVP query.',
+    'KEY': 'Key to validate.',
+    'MAX_AGE': 'Number of days a key should be considered valid.',
+    'SECRET': 'Secret to use for validation.',
+}
+
+REQUIRED_ARGS = [
+    'DB_HOST', 'DB_NAME', 'DB_PASS', 'DB_PORT', 'DB_USER', 'KEY', 'MAX_AGE',
+    'SECRET', 
+]
+
+
+def main(args):
+    key = validate_key.main(args)
+    if key.get('valid', False):
+        connection = psycopg2.connect(
+            host=args.DB_HOST,
+            port=args.DB_PORT,
+            user=args.DB_USER,
+            password=args.DB_PASS,
+            database=args.DB_NAME
+        )
+        cursor = connection.cursor(
+            cursor_factory=psycopg2.extras.RealDictCursor
+        )
+        query = """
+        SELECT
+            u.email, u.first_name, u.middle_name, u.last_name, u.state, u.city,
+            u.zip, MIN(a.created_at) AS action_datetime, MAX(s.role) AS role
+        FROM %s.core_user u
+        JOIN %s.events_eventsignup s ON s.user_id = u.id
+        JOIN %s.events_event e ON e.id = s.event_id
+        JOIN %s.events_campaign c ON c.id = e.campaign_id
+        LEFT JOIN %s.core_action a ON (
+            a.page_id = s.page_id
+            AND a.user_id = u.id
+        )
+        WHERE c.name = %s
+        %s
+        AND a.source = %s
+        GROUP BY 1,2,3,4,5,6,7""" % (
+                    args.DB_SCHEMA,
+                    args.DB_SCHEMA,
+                    args.DB_SCHEMA,
+                    args.DB_SCHEMA,
+                    args.DB_SCHEMA,
+                    '%s',
+                    args.EXTRA_WHERE,
+                    '%s'
+        )
+        cursor.execute(query, (key.get('campaign', ''), key.get('source', '')))
```

Comment:
Should `source` and/or `campaign` ever be allowed to be empty?  If so, can we bail here (either by assuming `key['campaign']` rather than with .get() or by bailing above.

Reply:
No, they can't be empty. That would fail key validation, which would prevent it from getting here. So it already bails before this if they're empty. Having defaults is just a general defensive coding practice to avoid key errors, which I'm pretty sure I picked up from you.